### PR TITLE
BL-5701 content with feature toggle (DO NOT MERGE TO MASTER)

### PR DESCRIPTION
**Pull request by mathieuhendey**
This PR adds a feature toggle check for SMS, and new content for SMSes that is only shown if the feature toggle is on.

### CODE REVIEW

File: webapp/src/main/java/uk/gov/dvsa/motr/web/cookie/MotrSession.java [hunk 1]
```diff
@@ -1,29 +1,41 @@
...
 
     private boolean shouldClearCookies;
+    private boolean allowedOnChannelSelectionPage;
+    private boolean allowedOnPhoneNumberEntryPage;
 
-    public MotrSession() {
+    public MotrSession(@SystemVariableParam(FEATURE_TOGGLE_SMS) Boolean featureToggleSms) {
 
         this.attributes = new HashMap<>();
+        this.allowedOnChannelSelectionPage = featureToggleSms;
```
Comment (aaronlong) [inline webapp/src/main/java/uk/gov/dvsa/motr/web/cookie/MotrSession.java]: Could the variables allowedOnChannelSelectionPage and allowedOnPhoneNumberEntryPage be removed.  Replacing them with the featureToggleSms variable.

File: webapp/src/main/java/uk/gov/dvsa/motr/web/resource/ChannelSelectionResource.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        return Response.ok(renderer.render(CHANNEL_SELECTION_TEMPLATE, modelMap)).build();
+    }
+
+    @POST
+    public Response channelSelectionPagePost(@FormParam("selectChannel") String formParamChannel) throws Exception {
+
+        ChannelSelectionValidator channelSelectionValidator = new ChannelSelectionValidator();
+
+        if (channelSelectionValidator.isValid(formParamChannel)) {
+            if (formParamChannel.equals("email")) {
```
Comment (aaronlong) [inline webapp/src/main/java/uk/gov/dvsa/motr/web/resource/ChannelSelectionResource.java]: For all of the "email" and "text" channel hardcoded text should we use a variable like TEXT_CHANNEL in motrSession

File: webapp/src/main/java/uk/gov/dvsa/motr/web/resource/EmailResource.java [hunk 3]
```diff
@@ -52,10 +52,16 @@ public Response emailPage() throws Exception {
             return redirect("/");
         }
 
+        if (!motrSession.visitingFromReviewPage()) {
+            motrSession.setVisitingFromContactEntry(true);
+        }
+
```
Comment (kevinc-kainos) [inline webapp/src/main/java/uk/gov/dvsa/motr/web/resource/EmailResource.java]: do we care about setting the the contrary case i.e. setVisitingFromContactEntry(false) when motrSession.visitingFromReviewPage() is true?
if we do then we could make it a single statement
motrSession.setVisitingFromContactEntry(!motrSession.visitingFromReviewPage())

File: selenium/src/test/java/uk/gov/dvsa/motr/helper/MotReminder.java [hunk 4]
```diff
@@ -26,6 +27,14 @@ public ReviewPage enterReminderDetails(String vrm, String email) {
         return emailPage.enterEmailAddress(email);
     }
 
+    public ReviewPage enterReminderDetailsSmsToggleOn(String vrm, String email) {
```
Comment (aaronlong) [inline selenium/src/test/java/uk/gov/dvsa/motr/helper/MotReminder.java]: Should probably be more clear that this method signs up for an email reminder from the method name :)

File: webapp/src/main/java/uk/gov/dvsa/motr/web/resource/ChannelSelectionResource.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        this.motrSession = motrSession;
+        this.renderer = renderer;
+        this.dataLayerHelper = new DataLayerHelper();
+    }
+
+    @GET
+    public Response channelSelectionPageGet() throws Exception {
+
+        if (!motrSession.isAllowedOnChannelSelectionPage()) {
+            return redirect("/");
```
Comment (aaronlong) [inline webapp/src/main/java/uk/gov/dvsa/motr/web/resource/ChannelSelectionResource.java]: Add unit test to check the redirection occours :)

File: webapp/src/main/java/uk/gov/dvsa/motr/web/resource/ChannelSelectionResource.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+        return Response.ok(renderer.render(CHANNEL_SELECTION_TEMPLATE, modelMap)).build();
+    }
+
+    @POST
+    public Response channelSelectionPagePost(@FormParam("selectChannel") String formParamChannel) throws Exception {
+
+        ChannelSelectionValidator channelSelectionValidator = new ChannelSelectionValidator();
+
+        if (channelSelectionValidator.isValid(formParamChannel)) {
```
Comment (aaronlong) [inline webapp/src/main/java/uk/gov/dvsa/motr/web/resource/ChannelSelectionResource.java]: Add unit test to ensure we cover what happens if the channel selected is not valid

File: webapp/src/main/java/uk/gov/dvsa/motr/web/resource/HomepageResource.java [hunk 7]
```diff
@@ -34,6 +39,9 @@ public HomepageResource(
     public String homePage() throws Exception {
 
         this.motrSession.setShouldClearCookies(true);
-        return renderer.render("home", emptyMap());
+        Map<String, Object> values = new HashMap<>();
```
Comment (aaronlong) [inline webapp/src/main/java/uk/gov/dvsa/motr/web/resource/HomepageResource.java]: Should we call this modelMap? :)

File: webapp/src/main/java/uk/gov/dvsa/motr/web/resource/PhoneNumberResource.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        this.motrSession = motrSession;
+        this.renderer = renderer;
+        this.dataLayerHelper = new DataLayerHelper();
+    }
+
+    @GET
+    public Response phoneNumberPageGet() throws Exception {
+
+        if (!motrSession.isAllowedOnPhoneNumberEntryPage()) {
+            return redirect("/");
```
Comment (aaronlong) [inline webapp/src/main/java/uk/gov/dvsa/motr/web/resource/PhoneNumberResource.java]: Add test for this :)

File: webapp/src/main/java/uk/gov/dvsa/motr/web/resource/PhoneNumberResource.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+        return Response.ok(renderer.render(PHONE_NUMBER_TEMPLATE, modelMap)).build();
+    }
+
+    @POST
+    public Response phoneNumberPagePost(@FormParam("phoneNumber") String phoneNumber) throws Exception {
+
+        PhoneNumberValidator validator = new PhoneNumberValidator();
+
+        if (validator.isValid(phoneNumber)) {
```
Comment (aaronlong) [inline webapp/src/main/java/uk/gov/dvsa/motr/web/resource/PhoneNumberResource.java]: Add test for if not valid

File: webapp/src/test/java/uk/gov/dvsa/motr/web/validator/ChannelSelectionValidatorTest.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
+package uk.gov.dvsa.motr.web.validator;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class ChannelSelectionValidatorTest {
```
Comment (aaronlong) [inline webapp/src/test/java/uk/gov/dvsa/motr/web/validator/ChannelSelectionValidatorTest.java]: Could we improve these tests by asserting that the message is also correct? :)

File: webapp/src/test/java/uk/gov/dvsa/motr/web/validator/PhoneNumberValidatorTest.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
+package uk.gov.dvsa.motr.web.validator;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class PhoneNumberValidatorTest {
```
Comment (aaronlong) [inline webapp/src/test/java/uk/gov/dvsa/motr/web/validator/PhoneNumberValidatorTest.java]: Assert Mesages :)

File: webapp/src/main/java/uk/gov/dvsa/motr/web/cookie/MotrSession.java (hunk 1 above)
Comment (mathieuhendey) [reply webapp/src/main/java/uk/gov/dvsa/motr/web/cookie/MotrSession.java]: I just added them for clarity, I think having them makes the code clearer plus it doesn't cost anything really. Makes it clearer for people looking at this code later on

Comment (aaronlong) [reply webapp/src/main/java/uk/gov/dvsa/motr/web/cookie/MotrSession.java]: Thanks, happy to leave it as it stands :)

**Pull request closed without merging**